### ISSUE TITLE
octave segment faults in parallel matrix multiplication 

### ISSUE THREAD

**Issue by sunilshah** [Homebrew/homebrew-core#35924]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [ x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
- [ ] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x ] ran `brew update` and can still reproduce the problem?
- [x ] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew install` commands)

octave segment faults in pararrayfun in matrix multiplication when the dimension exceeds a modest number, 4096 or 8192 depending upon the Mac hardware.  It appears that the script building the parallel package for octave may be using a library that does not support large enough matrix dimensions. This appears to happen only for matrix multiplication so far. 
 ````
function err=par_test2(in1) 
% 
%    Test this using pararrayfun(nproc,@(jj) 
par_test2(in1),1:4,"uniformoutput",false); 
a=randn(in1,2); 
b=randn(2,1); 
r1=a(1:in1,1)*b(1)+a(1:in1,2)*b(2); 
%  does not work for in1 =4097, works for in1=4096 
  r2=a*b; 
err=norm(r1-r2); 
endfunction 

````

````

octave:3> pararrayfun(nproc,@(jj) par_test2(4097),1:4,"uniformoutput",false); 
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
fatal: caught signal Segmentation fault: 11 -- stopping myself...
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
warning: parcellfun: premature exit due to closed pipe
warning: called from
    parcellfun at line 350 column 11
    pararrayfun at line 85 column 26

warning: broken pipe
warning: broken pipe
warning: broken pipe
warning: broken pipe
octave:4> pararrayfun(nproc,@(jj) par_test2(4096),1:4,"uniformoutput",false); 
parcellfun: 4/4 jobs done
octave:5> par_test2(4097)
ans = 0

````
I am using octave 4.4.1 on mac OS 10.13.6, installed using Homewbrew (octave 
version 4.4.1_4).  It is installed for cli (non-gui). 

$ brew info octave
octave: stable 4.4.1 (bottled), HEAD
High-level interpreted language for numerical computing
https://www.gnu.org/software/octave/index.html
/usr/local/Cellar/octave/4.4.1_2 (2,187 files, 48.5MB)
  Poured from bottle on 2018-10-21 at 16:34:09
/usr/local/Cellar/octave/4.4.1_3 (2,187 files, 48.5MB)
  Poured from bottle on 2018-11-28 at 11:26:42
/usr/local/Cellar/octave/4.4.1_4 (2,187 files, 48.5MB) *
  Poured from bottle on 2018-12-24 at 11:56:44
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/octave.rb
==> Dependencies
Build: gnu-sed ✔, pkg-config ✔
Required: arpack ✔, epstool ✔, fftw ✔, fig2dev ✔, fltk ✔, fontconfig ✔, freetype ✔, gcc ✔, ghostscript ✔, gl2ps ✔, glpk ✔, gnuplot ✘, graphicsmagick ✔, hdf5 ✔, libsndfile ✔, libtool ✔, pcre ✔, portaudio ✔, pstoedit ✔, qhull ✔, qrupdate ✔, readline ✔, suite-sparse ✔, sundials ✔, texinfo ✔, veclibfort ✔
Optional: qt ✔
==> Requirements
Build: java >= 1.6 ✔
==> Options
--with-qt
	Build with qt support
--HEAD
	Install HEAD version
==> Analytics
install: 7,262 (30 days), 23,254 (90 days), 100,326 (365 days)
install_on_request: 6,539 (30 days), 21,053 (90 days), 86,929 (365 days)
build_error: 0 (30 days)

On Ubuntu 16.04 LTS, octave 4.4.2 built using Mike Miller's ppa does not have this problem. Tested for dimension up to 1e8. 

I see [bug #49532]  on octave-forge with a similar problem in 2016.  

Is there a way to look at the compilation and linking flags for octave parallel package installation?  Do these flags get set through the homebrew formula?  

If there I should post this problem elsewhere, please advise.

**Comment by fxcoudert** [comment]
Here are the configure flags we pass:

```
    system "./configure", "--prefix=#{prefix}",
                          "--disable-dependency-tracking",
                          "--disable-silent-rules",
                          "--enable-link-all-dependencies",
                          "--enable-shared",
                          "--disable-static",
                          "--with-hdf5-includedir=#{Formula["hdf5"].opt_include}",
                          "--with-hdf5-libdir=#{Formula["hdf5"].opt_lib}",
                          "--with-x=no",
                          "--with-blas=-L#{Formula["veclibfort"].opt_lib} -lvecLibFort",
                          "--with-portaudio",
                          "--with-sndfile",
```

**Comment by fxcoudert** [comment]
Cannot reproduce, your code has syntax error.

```
octave:1> function err=par_test2(in1)
> %
> % Test this using pararrayfun(nproc,@(jj)
> par_test2(in1),1:4,"uniformoutput",false);
parse error:

  syntax error
```

As the issue template states, we also need the output of `brew doctor` and `brew config`.

**Comment by sunilshah** [comment]
Hi @fxcoudert 

Thanks for the response.  This blas flag may be the culprit.  
````
"--with-blas=-L#{Formula["veclibfort"].opt_lib} -lvecLibFort",

````
 To reproduce the problem in octave you need the definition of the function 
par_test2 in a file par_test2.m in the local directory (or in the load path).  In my original post, these lines  in the file par_test2.m

```
function par_test2(in2)
%
% Test this using pararrayfun(nproc,@(jj)
par_test2(in1),1:4,"uniformoutput",false);
a=randn(in1,2);
b=randn(2,1);
r1=a(1:in1,1)*b(1)+a(1:in1,2)b(2);
% does not work for in1 =4097, works for in1=4096
r2=ab;
err=norm(r1-r2);     
endfunction  
````

After invoking octave, you need to install parallel package from octave-forge if it is not installed and load it.

````
octave:1>pkg install -forge parallel
octave:2>pkg load parallel
````
Then the rest of the commands should work as in OP.
````
octave:3> pararrayfun(nproc,@(jj) par_test2(4097),1:4,"uniformoutput",false);
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
fatal: caught signal Segmentation fault: 11 -- stopping myself...
parcellfun: 0/4 jobs donefatal: caught signal Segmentation fault: 11 -- stopping myself...
warning: parcellfun: premature exit due to closed pipe
warning: called from
parcellfun at line 350 column 11
pararrayfun at line 85 column 26

warning: broken pipe
warning: broken pipe
warning: broken pipe
warning: broken pipe
octave:4> pararrayfun(nproc,@(jj) par_test2(4096),1:4,"uniformoutput",false);
parcellfun: 4/4 jobs done
octave:5> par_test2(4097)
ans = 0

````
Here are the requested brew command outputs
````
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: The following directories do not exist:
/usr/local/sbin

You should create these directories and change their ownership to your account.
  sudo mkdir -p /usr/local/sbin
  sudo chown -R $(whoami) /usr/local/sbin

Warning: You have unlinked kegs in your Cellar
Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@2
  qt@5.7

$ brew config
HOMEBREW_VERSION: 1.9.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 241ee989bb6a917cb6000a1aa06aa341c465a1b3
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: eda6e4eb4848beb52f64e7e94d502a435e5c6ef4
Core tap last commit: 16 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_LOGS: /Users/sunil/Library/Logs/Homebrew
CPU: quad-core 64-bit sandybridge
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1000
Git: 2.17.2 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 10.0.2
macOS: 10.13.6-x86_64
CLT: 10.1.0.0.1.1539992718
Xcode: 10.1
XQuartz: 2.7.11 => /opt/X11

````

**Comment by sunilshah** [comment]
Hi @fxcoudert 

Were you able to reproduce the segment fault? I did find a typo in the code which I corrected today.

 In [bug #49532] on octave-forge, Olaf Till conjectures that use of fork () may be the cause of segment fault. 

To test that hypothesis, I tested python 3.7 multiprocessing Pool implementation to see if it has any issues with large matrix multiplication.  It does not have the  problem even at dimension of 5000 x 5000.  I  did declare
      multiprocessing.set_start_method('fork')
So, use of fork() under Mac OS does not affect large parallel matrix multiply in Python 3.7.
